### PR TITLE
Support identity-provider-config for EKS

### PR DESCRIPTION
**Pull request by nnbu**
This change adds support for configuring identity-provider-config to an EKS cluster.

Associated apis: https://docs.aws.amazon.com/eks/latest/APIReference/API_AssociateIdentityProviderConfig.html , https://docs.aws.amazon.com/eks/latest/APIReference/API_DescribeIdentityProviderConfig.html , https://docs.aws.amazon.com/eks/latest/APIReference/API_DisassociateIdentityProviderConfig.html

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: apis/v1alpha1/identity_provider_config.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+)
+
+// IdentityProviderConfigSpec defines the desired state of IdentityProviderConfig.
+//
+// An object representing an identity provider configuration.
+type IdentityProviderConfigSpec struct {
+
+	// A unique, case-sensitive identifier that you provide to ensure the idempotency
+	// of the request.
+	ClientRequestToken *string `json:"clientRequestToken,omitempty"`
```
Comment (a-hilaly) [inline apis/v1alpha1/identity_provider_config.go:30]: I know some other resources still have this field.. ideally we do not want to include it in the spec.

File: config/crd/common/bases/services.k8s.aws_fieldexports.yaml [hunk 2]
```diff
@@ -3,7 +3,8 @@ apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
   annotations:
-    controller-gen.kubebuilder.io/version: v0.14.0
+    controller-gen.kubebuilder.io/version: v0.9.2
```
Comment (a-hilaly) [inline config/crd/common/bases/services.k8s.aws_fieldexports.yaml:6]: Can you please bump controller-tools to v0.14.0?

File: helm/crds/services.k8s.aws_adoptedresources.yaml [hunk 3]
```diff
@@ -183,12 +183,12 @@ spec:
                             name:
                               description: |-
                                 Name of the referent.
-                                More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names#names
+                                More info: http://kubernetes.io/docs/user-guide/identifiers#names
```
Comment (a-hilaly) [inline helm/crds/services.k8s.aws_adoptedresources.yaml:186]: I think you need to `git pull --rebase` the runtime repository

File: pkg/resource/addon/sdk.go [hunk 4]
```diff
@@ -457,17 +457,6 @@ func (rm *resourceManager) sdkUpdate(
...
-		ko.Status.CreatedAt = &metav1.Time{*resp.Update.CreatedAt}
-	} else {
-		ko.Status.CreatedAt = nil
-	}
-	if resp.Update.Status != nil {
-		ko.Status.Status = resp.Update.Status
-	} else {
-		ko.Status.Status = nil
-	}
-
```
Comment (a-hilaly) [inline pkg/resource/addon/sdk.go:470]: Mmmmm is there a reason why we're getting rid of those?

File: apis/v1alpha1/generator.yaml [hunk 5]
```diff
@@ -433,6 +467,8 @@ resources:
 ignore:
   resource_names:
   - EksAnywhereSubscription
+  shape_names:
+    - Update
```
Comment (a-hilaly) [inline apis/v1alpha1/generator.yaml:471]: Is this related to IdentityProviderConfig or Addons?

File: templates/hooks/identity_provider_config/sdk_create_post_set_output.go.tpl
Comment (a-hilaly) [inline templates/hooks/identity_provider_config/sdk_create_post_set_output.go.tpl:1]: 👍

File: templates/hooks/identity_provider_config/sdk_delete_post_build_request.go.tpl [hunk 6]
```diff
@@ -0,0 +1,6 @@
+	identityProviderConfigType := "oidc"
```
Comment (a-hilaly) [inline templates/hooks/identity_provider_config/sdk_delete_post_build_request.go.tpl:1]: Can we make this variable a constant in the hooks.go file?

File: test/e2e/resources/identity_provider_config.yaml [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+spec:
+  clusterName: $CLUSTER_NAME
+  oidc:
+    clientID: client
+    identityProviderConfigName: $IDP_NAME
+    issuerURL: https://some-url.com/
+    usernameClaim: user
+    groupsClaim: group
+  tags:
+    tkey: tval
```
Comment (a-hilaly) [inline test/e2e/resources/identity_provider_config.yaml:14]: Can we also support tag updates? I think we already have most of the logic in `pkg/util`

**Review by a-hilaly — COMMENTED** [review]
Great stuff @nnbu ! Thank you! I left a few comments and questions in line

File: apis/v1alpha1/generator.yaml (hunk 5 above)
Comment (nnbu) [reply apis/v1alpha1/generator.yaml:471]: This is related to identityProviderConfig.
[Associate](https://docs.aws.amazon.com/eks/latest/APIReference/API_AssociateIdentityProviderConfig.html) api returns `update` type. But [describe](https://docs.aws.amazon.com/cli/latest/reference/eks/describe-identity-provider-config.html) does not return that. So, essentially `update` turns out to be incomplete field which cant be updated once `associate` is done. e.g. `associate` sets one of the fields in `update` to be `inProgress`. But since `describe` does not have this, this field is never set to `active`. 

This `inPogress` string is not even part of statuses returned by `describe` api. ref: https://https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/eks/client/describe_identity_provider_config.html#describe-identity-provider-config

I did not find this to be useful to retain.

File: pkg/resource/addon/sdk.go (hunk 4 above)
Comment (nnbu) [reply pkg/resource/addon/sdk.go:470]: This got deleted since I removed `Update` shape in generator.yaml. I explained the reason in other comment. But I did not anticipate , it would delete the shape from `addon` as well. Should I keep `Update`?

File: apis/v1alpha1/identity_provider_config.go (hunk 1 above)
Comment (nnbu) [reply apis/v1alpha1/identity_provider_config.go]: done

File: config/crd/common/bases/services.k8s.aws_fieldexports.yaml (hunk 2 above)
Comment (nnbu) [reply config/crd/common/bases/services.k8s.aws_fieldexports.yaml]: done

File: helm/crds/services.k8s.aws_adoptedresources.yaml (hunk 3 above)
Comment (nnbu) [reply helm/crds/services.k8s.aws_adoptedresources.yaml]: done

File: templates/hooks/identity_provider_config/sdk_delete_post_build_request.go.tpl (hunk 6 above)
Comment (nnbu) [reply templates/hooks/identity_provider_config/sdk_delete_post_build_request.go.tpl]: done

File: test/e2e/resources/identity_provider_config.yaml (hunk 7 above)
Comment (nnbu) [reply test/e2e/resources/identity_provider_config.yaml:14]: [tagging resource requires arn](https://docs.aws.amazon.com/cli/latest/reference/eks/tag-resource.html). But identity provider config does not have arn associated with it. Even in the UI, there is no way to update its tags. So, I don't think, we can support tag updates.

File: pkg/resource/addon/sdk.go (hunk 4 above)
Comment (nnbu) [reply pkg/resource/addon/sdk.go]: I updated generator.yaml and added `  - AssociateIdentityProviderConfigOutput.Update`. This will only remove `update` field from identity-provider-config. and wont touch `Addon`

File: apis/v1alpha1/generator.yaml (hunk 5 above)
Comment (nnbu) [reply apis/v1alpha1/generator.yaml]: I updated generator.yaml and added   - AssociateIdentityProviderConfigOutput.Update. This will only remove update field from identity-provider-config. and wont touch Addon

File: test/e2e/resources/identity_provider_config.yaml (hunk 7 above)
Comment (nnbu) [reply test/e2e/resources/identity_provider_config.yaml:14]: I have made `tags` field immutable now.

**Review by a-hilaly — APPROVED** [review]
Neat and great work on this @nnbu - thank you so much!
/lgtm
